spec/consensus/light-client.md

```diff
@@ -1,318 +1,577 @@
+                              trustThreshold TrustThreshold,
+                              trustingPeriod Duration,
+                              clockDrift Duration,
+                              now Time) (error, TrustedState) {
+
+    trustedHeader = trustedState.SignedHeader.Header
+    assert trustedHeader.Height < untrustedHeight
+
+    untrustedSh, error := Commit(untrustedHeight)
+    if error != nil return (error, trustedState)
+
+    untrustedHeader = untrustedSh.Header
+    assert trustedHeader.Time < untrustedHeader.Time
+
+    // note that we pass now during the recursive calls. This is fine as
+    // all other untrusted headers we download during recursion will be
+    // for a smaller heights, and therefore should happen before.
+    assert untrustedHeader.Time < now + clockDrift
+
+    untrustedVs, error := Validators(untrustedHeight)
+    if error != nil return (error, trustedState)
+
+    untrustedNextVs, error := Validators(untrustedHeight + 1)
+    if error != nil return (error, trustedState)
+
+    error = verifySingle(
+             trustedState,
+             untrustedSh,
+             untrustedVs,
+             untrustedNextVs,
+             trustThreshold)
+
+    if fatalError(error) return (error, trustedState)
+
+    if error == nil {
+        // the untrusted header is now trusted.
+        newTrustedState = TrustedState(untrustedSh, untrustedNextVs)
+        return (nil, newTrustedState)
+    }
 
-- the lite client must retry to retrieve correct headers from another full node
-  - it picks a new full node
-  - it restarts _Bisection_
-  - there might be optimizations; a lite client may not need to call _Commit(k)_, for a height _k_ for which it already has a signed header it trusts.
-  - how to make sure that a lite client can communicate with a correct full node will be the focus of a separate document (recall Issue 3 from "Context of this document").
+    // at this point in time we need to do bisection
+    pivotHeight := ceil((trustedHeader.Height + untrustedHeight) / 2)
+
+    error, newTrustedState = VerifyAndUpdateBisection(trustedState,
+                                                      pivotHeight,
+                                                      trustThreshold,
+                                                      trustingPeriod,
+                                                      clockDrift,
+                                                      now)
+    if error != nil return (error, trustedState)
+
+    error, newTrustedState = verifyAndUpdateBisection(newTrustedState,
+                                                      untrustedHeight,
+                                                      trustThreshold,
+                                                      trustingPeriod,
+                                                      clockDrift,
+                                                      now)
+    if error != nil return (error, trustedState)
```

Comment:
If the second call to bisection failed, shouldn't we return the trusted state from the first call ? Or is the idea to only return the trusted state we started with or the one for the requested height, and nothing in between?

Reply:
We could potentially return intermediate trusted state with an error signalling that it does not corresponds to the untrustedHeight. In this case the guarantees we provide are that in case we return with no error (error = nil) you get trusted state that corresponds to the asked height (untrustedHeight); otherwise the function returns some trusted state that is at least as recent as trustedState.